### ISSUE TITLE
Running QGIS instance causes Windows installer to silently fail for minor version upgrades

### ISSUE THREAD

**Issue by AlisterH** [qgis/QGIS#31369]
Running 3.8.1 on Windows 10, it tells me I can upgrade to 3.8.2
I download and install 3.8.2 - it uninstalls 3.8.1, and then claims to install 3.8.2.  The start menu shortcut claims to be 3.8.2, but when I run it, the about dialog claims to be 3.8.1, and it still informs me that I can upgrade to 3.8.2.
Perhaps Windows is playing an unusually high tech trick on me, but more likely the 3.8.2 installer packages the wrong build!
I have not tried the OSGeo4W installer (I get in trouble for running it, because the retards at Trend Micro like to classify it as malware), or the 32 bit installer.

**Comment by roya0045** [comment]
Or might it be a metadata error that was not changed? IE the build version Id?

**Comment by AlisterH** [comment]
Don't think so.
This is the commit that is tagged 3.8.2: https://github.com/qgis/QGIS/commit/4470baa1a3369bce50d70f32f10b2326c7324bab
This is the commit that is tagged 3.8.1: https://github.com/qgis/QGIS/commit/dcd95cc648a5c3091866fbe808d4c315a2482bb3

**Comment by jef-n** [comment]
The 3.8.2 64bit installer is fine - checking 32bit next…

**Comment by AlisterH** [comment]
Don't bother, I've definitely only used the 64bit installer.

**Comment by AlisterH** [comment]
OK, further testing:
Knowing that this was reproducable on my work machine, where I had tried it three times, I found a machine elsewhere that has never had any QGIS version installed.  I downloaded and installed 3.8.2, and you are right, it claims to be 3.8.2.  
I downloaded and installed 3.8.1, and it claims to be 3.8.1.  
I left QGIS running in case that made a difference, and installed 3.8.2 again.  Strangely,
the uninstaller seemed to complete successfully, instead of stopping and complaining that the files were in use, as I'm sure it did until quite recently.  I guess Microsoft have finally realised the problems that people experience because of programs not releasing file locks on Windows.  Now when I start QGIS (even after closing it) it still claims to be 3.8.1, and tells me an upgrade is available, as per my original report.  
I will try rebooting now to see if that makes a difference.

**Comment by AlisterH** [comment]
After a reboot it still claims to be 3.8.1
Also note that when I originally tested I actually checked that I didn't have QGIS 3.8.1 open before running the 3.8.2 installer, so if the problem is related to file locks then you can't workaround it just by telling the user to close QGIS before running the installer.

**Comment by AlisterH** [comment]
Trying again on this machine to close QGIS before running the installer, it was successful.
I guess that means the problem is caused by file locking.  If this is essentially a Windows bug I guess these could be a few possible "solutions":
- allow installing minor versions alongside each other i.e. install 3.8.2 to a 3.8.2 folder, not just a 3.8 folder.  But it would probably need to be 3.8.2-1, to ensure it doesn't conflict with a 3.8.2-2
- tell users they have a retarded operating system and to check the QGIS version after installing, and reboot and try reinstalling if necessary.
- find a way for the installer to make sure it can actually install the new version, or at least detect whether it can install it, and advise to reboot first if not.

**Comment by AlisterH** [comment]
>  The start menu shortcut claims to be 3.8.2, but when I run it, the about dialog claims to be 3.8.1, and it still informs me that I can upgrade to 3.8.2.

When in this state, as well as the start menu shortcut claiming to be 3.8.2, when you run the installer again it falsely claims that 3.8.2 is already installed, and when the uninstaller finishes it says it has uninstalled 3.8.2 :)

Back on the original machine, I found a stray QGIS process in the task manager - after killing it the installer successfully installs 3.8.2.

**Comment by AlisterH** [comment]
I don't think this is as much of an issue now because you can install 3.22.1 and 3.22.2 side by side.
But it may still be an issue if there is a 3.22.2-1 and a 3.22.2-2

**Comment by AlisterH** [comment]
> I don't think this is as much of an issue now because you can install 3.22.1 and 3.22.2 side by side. But it may still be an issue if there is a 3.22.2-1 and a 3.22.2-2

I think technically there still could be a problem with -# versions, as I don't believe they install in a different folder.  But I think they also don't attempt to identify themselves, and updates to new -# versions aren't offered, so there wouldn't be users encountering this confusion. 
There probably isn't any point keeping this ticket open, and it may belong under osgeo4w, anyway.
